From fc4eb0ba1489e5ad5c5f5d2bd13cb7d0649586fd Mon Sep 17 00:00:00 2001
From: danfunk <daniel.h.funk@gmail.com>
Date: Wed, 3 Jan 2024 11:07:04 -0500
Subject: [PATCH 4/4] allow delete

---
 .../src/components/ProcessGroupForm.tsx       | 67 ++++++++++++++-----
 1 file changed, 49 insertions(+), 18 deletions(-)

diff --git a/spiffworkflow-frontend/src/components/ProcessGroupForm.tsx b/spiffworkflow-frontend/src/components/ProcessGroupForm.tsx
index fb7cc0fa..2d803b94 100644
--- a/spiffworkflow-frontend/src/components/ProcessGroupForm.tsx
+++ b/spiffworkflow-frontend/src/components/ProcessGroupForm.tsx
@@ -127,6 +127,30 @@ export default function ProcessGroupForm({
     updateProcessGroup(updateDict);
   };
 
+  const onDeleteMessage = (messageToDelete: Message) => {
+    // Remove message from process group
+    const newMessages = processGroup.messages?.filter((msg: Message) => {
+      return msg.id !== messageToDelete.id;
+    });
+
+    // Remove retrieval expressions from correlation properties related to message
+    const newCorrelationProperties: CorrelationProperty[] = JSON.parse(
+      JSON.stringify(processGroup.correlation_properties || [])
+    );
+    newCorrelationProperties.forEach((cp: CorrelationProperty) => {
+      // eslint-disable-next-line no-param-reassign
+      cp.retrieval_expressions = cp.retrieval_expressions.filter(
+        (re: RetrievalExpression) => {
+          return re.message_ref !== messageToDelete.id;
+        }
+      );
+    });
+    updateProcessGroup({
+      messages: newMessages,
+      correlation_properties: newCorrelationProperties,
+    });
+  };
+
   const formElements = () => {
     const textInputs = [
       <TextInput
@@ -178,7 +202,7 @@ export default function ProcessGroupForm({
 
   const [messageModelForModal, setMessageModelForModal] =
     useState<Message | null>(null);
-  const [correlationKeyForModal, setCorrelationKeyForModal] = useState<
+  const [correlationKeyForModal, setCorrelationKeyForMessageModal] = useState<
     CorrelationKey | undefined
   >(undefined);
 
@@ -193,11 +217,11 @@ export default function ProcessGroupForm({
           processGroup={processGroup}
           onClose={() => {
             setMessageModelForModal(null);
-            setCorrelationKeyForModal(undefined);
+            setCorrelationKeyForMessageModal(undefined);
           }}
           onSave={(updatedProcessGroup) => {
             setMessageModelForModal(null);
-            setCorrelationKeyForModal(undefined);
+            setCorrelationKeyForMessageModal(undefined);
             setProcessGroup(updatedProcessGroup);
           }}
         />
@@ -273,7 +297,7 @@ export default function ProcessGroupForm({
               hasIconOnly
               onClick={() => {
                 setMessageModelForModal(msg);
-                setCorrelationKeyForModal(correlationKey);
+                setCorrelationKeyForMessageModal(correlationKey);
               }}
             >
               Edit process group
@@ -284,8 +308,10 @@ export default function ProcessGroupForm({
               renderIcon={TrashCan}
               iconDescription="Delete Message"
               hasIconOnly
-              description={`Delete process group: ${processGroup.display_name}`}
-              onConfirmation={() => {}}
+              description={`Delete message: ${msg.id}`}
+              onConfirmation={() => {
+                onDeleteMessage(msg);
+              }}
               confirmButtonLabel="Delete"
             />
           </td>
@@ -317,14 +343,17 @@ export default function ProcessGroupForm({
       processGroup.correlation_keys &&
       processGroup.correlation_keys.length > 0
     ) {
-      items = processGroup.correlation_keys.map((ck: CorrelationKey) => {
-        const dataQA = `add-message-button-${ck.id}`;
+      // @ts-ignore
+      const keys = [undefined].concat(processGroup.correlation_keys);
+      items = keys.map((ck: CorrelationKey | undefined) => {
+        const id = ck?.id || 'uncorrelated';
+        const dataQA = `add-message-button-${id}`;
         return (
           <>
             <br />
             <br />
             <h3>
-              Correlation <b>{ck.id}</b>
+              <b>{id}</b> Messages
               <Button
                 kind="ghost"
                 data-qa={dataQA}
@@ -333,7 +362,7 @@ export default function ProcessGroupForm({
                 hasIconOnly
                 onClick={() => {
                   setMessageModelForModal({ id: '' });
-                  setCorrelationKeyForModal(ck);
+                  setCorrelationKeyForMessageModal(ck);
                 }}
               >
                 Edit process group
@@ -345,18 +374,20 @@ export default function ProcessGroupForm({
         );
       });
     }
-    const body = (
-      <>
-        <h3>Uncorrelated Messages</h3>
-        {messageDisplay()}
-        {items}
-      </>
-    );
     return (
       <div className="processGroupMessages">
         {messageEditModal()}
         <h2>Messages</h2>
-        {body}
+        {items}
+        <Button
+          kind="ghost"
+          data-qa="add-correlation-button"
+          renderIcon={Add}
+          iconDescription="Add Correlation"
+          hasIconOnly
+        >
+          Edit process group
+        </Button>
       </div>
     );
   };
-- 
2.25.1

